### PR TITLE
Add SystemGraph as a handle-based method for explicitly creating system dependency graphs

### PR DESCRIPTION
**Pull request by james7132**
# Objective
This PR adds `SystemGraph` as way to create system dependency graphs without explicit labeling.

## Motivation
It is currently difficult to force a strict system dependency graph. Currently the only way is manually set `label`, `before`, and/or `after` on each system. This can lead to some particularly boilerplate/non-ergonomic builder code that is difficult to read, harder to think about, and more prone to programmer error. An example of this kind of code can be found here: https://github.com/HouraiTeahouse/FantasyCresendoBevy/blob/master/src/match/mod.rs#L248.

Natural ordering invariants in game dev are common, and this should provide a simpler way to define those requirements. One notable case is the common simple sequential execution. Deterministic simulation is much easier to reason about with a single execution order, which should make it easier to create networked games that rely on determinism.

## Solution
This PR adds `SystemGraph`. It provides a handle based API for creating dependency graphs, and can be converted into SystemSets. It enforces ordering by generating `NodeId`s, a newtype around u32, as a globally unique system label, that is then used to label and order systems relative to each other.

This should be able to be used alongside "normal" SystemLabels to allow for ordering other systems relative any system in the SystemSet.

As for the choice to use AtomicU32, unless a developer is adding >2^32 systems to a stage,  there will be no collisions. If multiple Apps are set up and recycled in the same process (i.e. headless server environments spinning up game instances), the static AtomicU32 uses wrapping adds when generating IDs, so there is no risk of crashing from an overflow.

### CODE REVIEW

File: crates/bevy_ecs/src/schedule/system_set.rs
Comment (MinerSebas) [inline crates/bevy_ecs/src/schedule/system_set.rs:97]:
```suggestion
    /// If called, all systems added will run sequentially in insertion order.
```

File: crates/bevy_ecs/src/schedule/system_set.rs
Comment (mockersf) [inline crates/bevy_ecs/src/schedule/system_set.rs:170]: Why do you use an `AtomicU32` instead of just something local to this system set? How would I add to sequential system sets in parallel?

File: crates/bevy_ecs/src/schedule/system_set.rs
Comment (NathanSWard) [inline crates/bevy_ecs/src/schedule/system_set.rs:199]: Any reason to not just
```rust
#[derive(SystemLabel)]
```
instead of manually implementing it?

File: crates/bevy_ecs/src/schedule/system_set.rs
Comment (NathanSWard) [reply crates/bevy_ecs/src/schedule/system_set.rs:170]: I'm generally against globals (if they can be avoided) so if there is a way to localize this, I would prefer that. 
However, this is a small scope where if there wasn't a way to localize it, that would be ok.

File: crates/bevy_ecs/src/schedule/system_set.rs
Comment (NathanSWard) [inline crates/bevy_ecs/src/schedule/system_set.rs:9]:
```suggestion
use std::{fmt::Debug, hash::Hash, sync::atomic::{AtomicU32, Ordering}};
```

**Review by NathanSWard — CHANGES_REQUESTED** [review]

**Comment by NathanSWard** [comment]
It would be nice to add a test to ensue the is working as intended (it looks good to me, but tests always help)

File: crates/bevy_ecs/src/schedule/system_set.rs
Comment (james7132) [reply crates/bevy_ecs/src/schedule/system_set.rs:170]: The main reason is to generate a compact but effectively globally unique value.

I considered using UUIDs or just random u64s, but there is always a 1/2^n chance of a collision, and while the chance that a collision happens is astronomically small in both cases, a random collision may result seemingly random performance degradation.

This AtomicU32 is both smaller and guarantees that, unless a user adds 2^32 systems to a stage, there is zero chance of a collision. Most consumer machines, and most hosted VMs will likely OOM before that happens. Could try AtomicU16, but I can think of a few use cases to have tens of thousands of systems.

If we localize the u32 to a given instance, we need some point of reference such that other SystemSets are guaranteed not to have an overlap. I currently cannot think of an alternative that affords the same guarantees.

Comment (mockersf) [reply crates/bevy_ecs/src/schedule/system_set.rs:169]: Actually, couldn't you just declare a `SequenceId` type local to the method?

File: crates/bevy_ecs/src/schedule/system_set.rs
Comment (james7132) [reply crates/bevy_ecs/src/schedule/system_set.rs:198]: I get this error when using it:
```
use of undeclared crate or module `bevy_ecs`
```

Seems like the derive macro doesn't work within bevy_ecs itself.

File: crates/bevy_ecs/src/schedule/system_set.rs
Comment (james7132) [reply crates/bevy_ecs/src/schedule/system_set.rs:169]: As discussed on Discord, even if we did that, `SequenceId` would have the same type on every function call.

**Comment by mockersf** [comment]
bors try

**Comment by james7132** [comment]
> It would be nice to add a test to ensue the is working as intended (it looks good to me, but tests always help)

Added two tests for ensuring that labels were generated in sequence when sequential is enabled and are not present otherwise.

File: crates/bevy_ecs/src/schedule/system_set.rs
Comment (NathanSWard) [reply crates/bevy_ecs/src/schedule/system_set.rs:202]: Ugh, that's annoying :(

File: crates/bevy_ecs/src/schedule/system_set.rs
Comment (NathanSWard) [reply crates/bevy_ecs/src/schedule/system_set.rs:173]: Yea, I'm perfectly ok with the `AtomicU32` solution for now. 
We can't seem to find a better alternative so this should suffice imo.

**Review by NathanSWard — APPROVED** [review]
This looks good to me. 👍 
I'm still not the biggest fan of the global `AtomicU32`, however, I don't know of another way to implement this without risks of label collision, so the current impl should suffice :)

**Comment by james7132** [comment]
Reproducing some points from a followup discussion on Discord:

 * The end goal is to produce a distinct group of system DAGs 
 * Labels only exist at app build time, not during actual execution. 
 * Adding a large number of `bevy_ecs` provided labels may end up being noisier with tools that visualize the execution graph.
 * Ultimately it boils down to specifying dependencies in `ParallelSystemContainer::dependencies`. 
 * It's best not to have multiple confusing ways of specifying system execution dependencies (i.e. labels, `SystemSet::as_sequential`, etc.)
 
IMO, explicitly labeling can be quite error prone for common use cases, particularly with larger and more complex dependency graphs. However, labels still have their use in exposing dependencies to other crates. Likewise, with these points, it's likely that `SystemSet::as_sequential` is going to cut it for these kinds of use cases, and may end up just fragmenting the many ways to configure system dependencies.

The dependency graph seems very similar to how some big data pipeline services and frameworks like Apache BEAM and Google Cloud DataFlow build their DAGs. An example of one of these DAGs can be seen here: 
image

These frameworks often have a graph handle based implementation of defining them in code. Here's a possible example of how such an API might look:

```rust
let graph = SystemGraph::new();
let start_a = graph.root(sys_a.system().label("ENTRY_A"));
let start_b = graph.root(sys_b.system().label("ENTRY_B"));

let mid_c = start_a.then(sys_c.system());
let mid_d = start_a.then(sys_d.system());

(start_b, mid_c, mid_d)
  .join()
  .then(sys_e.system().label("EXIT_E"));

App::build().add_system_set(graph.into());
```

This should provide a general solution that enables every possible dependency graph without explicit labeling. An example that `SystemSet::as_sequential` can cover right now:
```rust
let graph = SystemGraph::new();
graph.root(sys_a.system())
  .then(sys_b.system())
  .then(sys_c.system())
  .then(sys_d.system())
  .then(sys_e.system());
App::build().add_system_set(graph.into());
```

A public interface can be relatively easily made building on top of what is in this PR using labels, and be switched to a label-less implementation later.

Would something like this need a RFC? As proposed here, this could be implemented as a utility ecosystem crate since it simply builds a SystemSet, but it would require first-party integration when migrating to a label-less implementation.

**Comment by james7132** [comment]
Pushed a tentative implementation of a SystemGraph API loosely based on the SystemSet::as_sequential implementation. This should allow creating SystemSets that as_sequential allowed, but also support both of the fan-in/fan-out topologies as well.

This implementation uses Arc/Mutex to ensure the resultant types are Send/Sync, but it's not completely necessary and can be done with Rc/RefCell instead.

**Comment by NathanSWard** [comment]
> This implementation uses Arc/Mutex to ensure the resultant types are Send/Sync, but it's not completely necessary and can be done with Rc/RefCell instead.

If we don't need Arc/Mutex and can get away with Rc/RefCell, we should imo :)

File: crates/bevy_ecs/src/schedule/system_graph.rs
Comment (NathanSWard) [inline crates/bevy_ecs/src/schedule/system_graph.rs:47]: I think we could probably improve on this API.
Instead of this, I'd want to do something like:

```rust
graph.root(sys_a.system()).then((sys_b.system(), sys_c.system()));
```
Then should be able to take 1 or more systems instead of having to store the variable and then call `then` twice (with-out chaining).

Also I could easily see a user getting confused on the difference between chaining `then` calls and then making a variable and calling `then` on the same thing twice.

File: crates/bevy_ecs/src/schedule/system_graph.rs
Comment (NathanSWard) [inline crates/bevy_ecs/src/schedule/system_graph.rs:3]: we should use `bevy_utils::HashMap`

File: crates/bevy_ecs/src/schedule/system_graph.rs
Comment (NathanSWard) [inline crates/bevy_ecs/src/schedule/system_graph.rs:7]: While I prefer the `Rc`/`RefCell` approach, if we do use a `Mutex` we should use `parking_lot::Mutex`

File: crates/bevy_ecs/src/schedule/system_graph.rs
Comment (NathanSWard) [inline crates/bevy_ecs/src/schedule/system_graph.rs:89]: Any specific reason we have this take a `&self`?
This probably jumps back to the Arc/Mutex thing I mentioned before. 
However, because this is a "builder"-like struct, it would make more sense for this to take `&mut self` or `self`

File: crates/bevy_ecs/src/schedule/system_graph.rs
Comment (NathanSWard) [inline crates/bevy_ecs/src/schedule/system_graph.rs:149]:
```suggestion
        self.graph.add_dependency(self.id, dst);
```
`add_dependent` is a single line function, it should probably just be inlined here.

File: crates/bevy_ecs/src/schedule/system_graph.rs
Comment (NathanSWard) [inline crates/bevy_ecs/src/schedule/system_graph.rs:159]: `join_into` is not very descriptive of what this is doing imo. 
Perhaps
`join_into_node`
`coalesce_system_nodes`
`merge_nodes`

**Review by NathanSWard — CHANGES_REQUESTED** [review]
I really like where this is headed 😄 
I think using graph-like syntax to define the dependencies of systems is really nice. 
Also, I also like the push to move away from a label-based API

File: crates/bevy_ecs/src/schedule/system_graph.rs
Comment (james7132) [reply crates/bevy_ecs/src/schedule/system_graph.rs:47]: I'm actually a bit more confused by that example. Does calling with a tuple of systems return the node representing the last system in the chain? Does it return a tuple and fan-out? Can I input a iterator and get an iterator of fan-out systems?

I think it's worth splitting that out into it's own function like `join_into` rather than having a single entry point.

File: crates/bevy_ecs/src/schedule/system_graph.rs
Comment (james7132) [reply crates/bevy_ecs/src/schedule/system_graph.rs:89]: It takes &self since it all the state is interior to it. I guess &mut self might make sense for enforcing users mark variables as mut, but it's not strictly necessary. Using self would not be wise either since you can have multiple roots in a given graph.

File: crates/bevy_ecs/src/schedule/system_graph.rs
Comment (james7132) [reply crates/bevy_ecs/src/schedule/system_graph.rs:159]: I used join because that's what the dataflow graph APIs used. It's close in both name and function to APIs like futures::join. 

Merge implies the existing nodes are consumed to make the the new one. Whereas it's actually just linking the new system to all of its dependencies. Not sure about that one.

File: crates/bevy_ecs/src/schedule/system_graph.rs
Comment (james7132) [reply crates/bevy_ecs/src/schedule/system_graph.rs:7]: Should have checked the dependencies, didn't want to assume it was available.

File: crates/bevy_ecs/src/schedule/system_graph.rs
Comment (james7132) [reply crates/bevy_ecs/src/schedule/system_graph.rs]: Added a `split_into` implementation that does just this.

**Comment by james7132** [comment]
Added a `split_into` utility function for simplifying fan-out topologies. Had to add a fix to bevy_ecs_macros to support tuple starts at something other than 0.

File: crates/bevy_ecs/macros/src/lib.rs
Comment (NathanSWard) [inline crates/bevy_ecs/macros/src/lib.rs:48]: What's the purpose of this change?
Wouldn't this break other uses of `all_tuples!`?

Comment (james7132) [reply crates/bevy_ecs/macros/src/lib.rs:48]: Without this change, the slicing on line 68 panics during compilation. The changed region here only generates the tuple types that are then filtered on line 68. Since SystemGroup is really only valid for tuples of size 2 and higher, it failed to compile without this change.

**Comment by bjorn3** [comment]
> As for the choice to use AtomicU32, unless a developer is adding >2^32 systems to a stage, there will be no collisions. If multiple Apps are set up and recycled in the same process (i.e. headless server environments spinning up game instances), the static AtomicU32 uses wrapping adds when generating IDs, so there is no risk of crashing from an overflow.

That will cause it to silently misbehave when adding >2^32 systems to a stage or if you add one system to one stage, 2^32 to another stage and then one to the first stage. Maybe you could have the counter local to each SystemGraph and panic on overflow. The label can then contain the local counter and a global SystemGraph counter.

**Comment by james7132** [comment]
Implemented the panicking add and atomic graph/node ID split. Ultimately (as in, not in this PR), however, it's likely desirable to not  use labels at all and just opt for writing the metadata to Parallel/ExclusiveSystemDescriptor.

File: crates/bevy_ecs/src/schedule/system_graph.rs
Comment (NathanSWard) [inline crates/bevy_ecs/src/schedule/system_graph.rs:188]: Nice work on getting the tuple api to work :)
However, why do we need to split the `then`/`split_into` APIs?

Ideally I'd want something like:
```rust
let root = SystemGraph::new().root(a.system()).then(b.system()).then(c.system());
```
this constructs a graph:
```
A -> B -> C
```

however I would also like to do:
```rust
let root = SystemGraph::new()
    .root(a.system())
    .then((b1.system(), b2.system()))
    .then(c.system());
```
which would construct a 
```
   / B1  \
A          C
   \ B2  /
```
e.g. the type that is returned from `then` should be able to either chain a `then` command where the previously added system(s) (`B1` and `B2`) are run in parallel and then the newly added `C` system is run after `B1` and `B2`

Optionally, then we should be able to break apart `B1` and `B2` and add independent chains e.g.
```rust
let (b1, b2) = SystemGraph::new()
    .root(a.system())
    .then((b1.system(), b2.system()));

b1.then(c1.system());
b2.then(c2.system());
```
which would produce
```
   / B1  -> C1
A          
   \ B2  -> C2
```

Perhaps this might be over engineering, but I like keeping the interface similar. 
Though optionally, perhps `split_into` should be named `fork` or something like that?

File: crates/bevy_ecs/src/schedule/system_graph.rs
Comment (NathanSWard) [inline crates/bevy_ecs/src/schedule/system_graph.rs:217]: What's the `[allow(non_snake_case)]` suppressing?

File: crates/bevy_ecs/src/schedule/system_graph.rs
Comment (NathanSWard) [reply crates/bevy_ecs/src/schedule/system_graph.rs:229]: perhaps then just `join`?

Comment (james7132) [reply crates/bevy_ecs/src/schedule/system_graph.rs:229]: Sounds good to me. Matches well with `fork`.

File: crates/bevy_ecs/src/schedule/system_graph.rs
Comment (james7132) [reply crates/bevy_ecs/src/schedule/system_graph.rs:217]: Line 219 destructures the tuple into variables using the tuple type parameters (T1, T2, T3...), which trigger a snake case warning. This suppresses it. If you look at the tuple implementation for SystemParam, it does the exact same thing.

File: crates/bevy_ecs/src/schedule/system_graph.rs
Comment (james7132) [reply crates/bevy_ecs/src/schedule/system_graph.rs:188]: I intentionally chose this to avoid potentially surprising behavior like the following:

```
let (b1, b2) = SystemGraph::new() 
  .root(a.system()) 
  .then((b1.system(), b2.system()))
  .then((c1.system(), c2.system()));
```

where it's unclear if it would produce something like what you have in the second example, or the following:

```
    / B1 -> C1
A          X
    \ B2 -> C2
```

I definitely like the matchup of `fork`/`join` though. I'll go ahead and rename the function.

Comment (james7132) [reply crates/bevy_ecs/src/schedule/system_graph.rs]: For the one -> many -> one case, I added implementations for joining the output of `fork` together. This is now valid:

```rust
let root = SystemGraph::new()
    .root(a.system())
    .fork((b1.system(), b2.system()))
    .join(c.system());
```

This should allow us to make any one to one, one to many, and many to one dependencies, but any many to many dependencies need to manually constructed.

**Review by NathanSWard — COMMENTED** [review]
This looks really good :)
Sorry to be a bother haha, but the last thing I'd add is we should provide an example in the `examples/` directory.
While the doc examples is a great start, actually seeing this in a runnable example would be very nice imo 😄

File: crates/bevy_ecs/src/schedule/system_graph.rs
Comment (NathanSWard) [inline crates/bevy_ecs/src/schedule/system_graph.rs:177]: Also, I forget what the API for `add_system_set` is, however we can probably update it to
```rust
fn add_system_set(set: impl Into<SystemSet>)
```
That way we don't have to call `graph.into()` when passing them as this parameter :)

**Review by NathanSWard — CHANGES_REQUESTED** [review]

**Review by Ratysz — COMMENTED** [review]
I quite like this, it's even better than what I had in mind whenever I mentioned "ergonomic APIs for simple cases" (like [here](https://github.com/bevyengine/bevy/discussions/2259#discussioncomment-807314), last paragraph).

I'll have to go over the graph implementation more thoroughly, but so far all of the potential kinks seem to have been addressed.

> Also, I also like the push to move away from a label-based API

I would put it as "build on top of the label API", like it was the intention all along. We finally have someone go and actually do it.

File: crates/bevy_ecs/macros/src/lib.rs
Comment (mockersf) [inline crates/bevy_ecs/macros/src/lib.rs:48]: wouldn't that be just `input.end`?

File: crates/bevy_ecs/src/schedule/system_graph.rs
Comment (mockersf) [inline crates/bevy_ecs/src/schedule/system_graph.rs:85]: probably wrong title

File: crates/bevy_ecs/src/schedule/system_graph.rs
Comment (mockersf) [inline crates/bevy_ecs/src/schedule/system_graph.rs:109]: wrong link?

**Comment by mockersf** [comment]
bors try

**Comment by mockersf** [comment]
Tests only test for system count in a graph and the presence of labels, it would be nice to have tests on the actual ordering.

Also I would love to see a debug view of a system graph in a followup pr

**Comment by NathanSWard** [comment]
bors try

**Review by NathanSWard — COMMENTED** [review]
This looks to be in a pretty good state imo :)
A few lingering questions that I'll just throw out there:
How would I express something like:
run A&B in parallel _then_ return C&D in parallel.
Perhaps a `with` function of something of the sort could be nice?
```rust
let a = SystemGraph::new().root(sys_a);
let first = a.with(sys_b, sys_c);
first.then(sys_d);
```

This would build something like:
```
A \
B - C
C /
```
A&B&C run in parallel _then_ D.

File: examples/ecs/system_graph.rs
Comment (Ratysz) [inline examples/ecs/system_graph.rs:18]:
```suggestion
            print_system("Sequential 3")
                .label("Sequential End"),
```

File: examples/ecs/system_graph.rs
Comment (Ratysz) [inline examples/ecs/system_graph.rs:24]:
```suggestion
        .root(print_system("Wide Start").after("Sequential End"))
```
Omitting `.system()` should be okay here now.

File: crates/bevy_ecs/src/schedule/system_graph.rs
Comment (Ratysz) [inline crates/bevy_ecs/src/schedule/system_graph.rs:216]:
```suggestion
    /// Creates a new node in the graph and adds the current node as its dependency.
```
Yeah, English language is weird. Pronouns don't get a posessive apostrophe (his, hers, its).

File: crates/bevy_ecs/src/schedule/system_graph.rs
Comment (Ratysz) [inline crates/bevy_ecs/src/schedule/system_graph.rs:300]: I think it's okay to use "must" here, seeing as this is an irrecoverable error.

File: crates/bevy_ecs/src/schedule/system_graph.rs
Comment (Ratysz) [inline crates/bevy_ecs/src/schedule/system_graph.rs:46]:
```suggestion
/// let (c, b, d) = graph.root(sys_a)
```

**Review by Ratysz — APPROVED** [review]
LGTM, as long as these nits are addressed.

**Comment by james7132** [comment]
> How would I express something like:
> run A&B in parallel _then_ return C&D in parallel.

IMO the API should avoid easy to misunderstand terms like `with` to convert into a type that overloads `then`.
I added `par_join` to SystemJoin that takes a SystemGroup and creates a many-to-many dependency mapping, though I'm open to other suggestions for how to name the public facing API though.

**Comment by Ratysz** [comment]
To me, `par_join()` sounds like the process of the joining itself is distributed across several threads, for some reason; I'd prefer something like `group_join()`.

Alternatively, it might be possible to make `join()` generic over system and system group, using an intermediary trait.

**Comment by NathanSWard** [comment]
> Alternatively, it might be possible to make join() generic over system and system group, using an intermediary trait.

I agree. Having a generic `join` would be best imo :)

File: crates/bevy_ecs/src/schedule/system_graph.rs
Comment (alice-i-cecile) [inline crates/bevy_ecs/src/schedule/system_graph.rs:206]: I would like some simple doc strings here, since this is pub.

File: crates/bevy_ecs/src/schedule/system_graph.rs
Comment (alice-i-cecile) [inline crates/bevy_ecs/src/schedule/system_graph.rs:213]: The ref-counting detail (where it points to the same underlying graph) is important to mention here; getting access to the same data without being handed a `&` is surprising (but handy!).

File: crates/bevy_ecs/src/schedule/system_graph.rs
Comment (alice-i-cecile) [inline crates/bevy_ecs/src/schedule/system_graph.rs:252]: Maybe `join_all`? `par_join` is not a very intuitive name for this to me.

File: crates/bevy_ecs/src/schedule/system_graph.rs
Comment (alice-i-cecile) [inline crates/bevy_ecs/src/schedule/system_graph.rs:169]: `dst` is not a clear variable name in context; I can't even guess what this is :)

**Review by alice-i-cecile — APPROVED** [review]

**Review by mockersf — APPROVED** [review]

**Review by NathanSWard — APPROVED** [review]
Overall I'm pretty happy with the PR.
Im not 100% convinced on the API names, however idk if I'll ever be haha 😆.
I guess we can let cart decide :)

**Comment by alice-i-cecile** [comment]
@james7132 when you get a chance can you leave a comment in #2373?

**Comment by james7132** [comment]
Done!

**Comment by james7132** [comment]
bors try

**Comment by james7132** [comment]
Made the main chunk of this PR available as an ecosystem crate in the form of [bevy_system_graph](https://crates.io/crates/bevy_system_graph). I do want some form of this as a first-party solution since this seems like something users generally need in a standardized way, but with stageless on the horizon I'll be closing this and opening a new PR once stageless has landed in some form.

The fixes to `all_tuples` macro, I'll split into a separate PR.

**Pull request closed without merging**
